Unified QEV Calculation #167

IanMendozaJaimes opened the pull request:
- Added new tables to store daily transactions and total points per day. Added a totals table
- Added QEV calculation
- Modified transaction points calculation for orgs and users
- Modified number of transactions validation for becoming a resident/citizen user or reputable/regenerative org
- Added tests

I also pushed a branch called migration/unified-qev-calculation, there I put a method called `migrateusers`, that action will copy the transactions for users and orgs and store them in the new tables using the new criteria to count transactions. The method doesn't clean the old transactions tables just in case.

## Deployment

- [x] ./scripts/seeds.js run accounts harvest history organization settings

- [x] Test accounts
- [x] Test harvest
- [x] Test history
- [x] Test organization

## Deploy Testnet
- [ ] Deploy 4 contracts to testnet
- [ ] ./scripts/seeds.js updatePermissions
- [ ] reset scheduler
- [ ] Run migration on testnet
- [ ] Ensure all values are as expected

## Deploy Mainnet
- [ ] Deploy 4 contracts
- [ ] ./scripts/seeds.js updatePermissions
- [ ] reset scheduler
- [ ] Run migration
- [ ] Ensure all values are as expected

juliolrmonteiro reviewed (approved):
Looks good to me, I would just try to merge tables `qev_table` and `transaction_points_table` in a single one, but this is not required.

n13 commented on src/seeds.history.cpp:
```diff
@@ -113,202 +126,236 @@ void history::trxentry(name from, name to, asset quantity) {
+    transaction.volume = quantity.amount;
+    transaction.qualifying_volume = std::min(transactions_cap, quantity.amount);
+    transaction.from_points = uint64_t(ceil(from_trx_multiplier * utils::get_rep_multiplier(to)));
+    transaction.to_points = uint64_t(ceil(to_trx_multiplier * utils::get_rep_multiplier(from)));
```
I have a question on this
- if "to" is an individual then transaction.to_points  should be == 0 (as individuals do not receive points on incoming transactions)

IanMendozaJaimes replied:
Oh yeah, that's right I am not setting them to 0, although the field "to" is not being used to calculate transaction points for individuals, but yeah I will change it.

n13 commented on src/seeds.history.cpp:
```diff
@@ -113,202 +126,236 @@ void history::trxentry(name from, name to, asset quantity) {
-      item.in = false;
-      item.quantity = quantity;
-      item.timestamp = eosio::current_time_point().sec_since_epoch();
+  double to_trx_multiplier = std::min(max_transaction_points_organizations, quantity.amount) / 10000.0;
```
Why is this divided by 10000?

Should this not be called "capped_amount" - seems like that's what it is the quantity, capped to the limit ?

IanMendozaJaimes replied:
I divided it by 10000 because I mimicked the way transaction points were originally calculated in harvest. Yes, capped_amount is a better name, I will change it

n13 replied:
Right so points are assigned on Seeds traded by the amounts are x10,000 ... makes sense.

n13 commented on src/seeds.history.cpp:
```diff
@@ -113,202 +126,236 @@ void history::trxentry(name from, name to, asset quantity) {
+      ft_itr -> from == from && ft_itr -> to == to) {
+
+    if (ft_itr -> volume < current_itr -> volume) {
+      current_itr = ft_itr;
```
current_itr is set to the smallest of the - potentially - 3 transactions

IanMendozaJaimes replied:
Yes, I you said below, it erases the smallest so it keeps the 2 greatest qualifying transactions. I did it that way because I didn't want to determine that in the `trxentry` action due to the reliability concerns, that's why it adds the transaction to the table and then in `savepoints` checks if it should be kept.

n13 commented on src/seeds.history.cpp:
```diff
@@ -113,202 +126,236 @@ void history::trxentry(name from, name to, asset quantity) {
+  if (count > max_number_transactions) {
+    from_points -= current_itr -> from_points;
+    to_points -= current_itr -> to_points;
+    qualifying_volume -= current_itr -> qualifying_volume;
```
So this could either end up in some positive values, or it could be all 0 if the new transaction is the smallest transaction

Then it deletes the smallest, so ends back up with 2 transactions

IanMendozaJaimes replied:
Yes, that's correct. But it also could be negative values, those are calculated in order to adjust the total points, so even if they become negative, the total transaction points calculated will never be negative. I realized that a transaction could have a grater volume, but not necessarily give more transaction points, I thought we had to keep the 2 transactions with the biggest volume, is that right?

n13 replied:
Yes on the 2tx

I'll trust that it works on the others, not sure how they can be negative ;)

n13 commented on src/seeds.history.cpp:
```diff
@@ -113,202 +126,236 @@ void history::trxentry(name from, name to, asset quantity) {
 
-  //print("orgtxpt for "+org.to_string());
+  save_from_metrics (from, from_points, qualifying_volume, day);
 
```
so save_from_metrics gets *additional* points - which could potentially be 0, but cannot be negative? Right?

IanMendozaJaimes replied:
Mmm nope, it adjusts the total transaction points for "from", so  from_points could be negative, the total transaction points won't be. Here the algorithm is preferring the transaction that moves more Seeds rather than the transaction that gives more points. Would you like me to change that?

n13 commented on test/harvest.test.js:
```diff
@@ -504,45 +528,48 @@ describe("harvest transaction score", async assert => {
   let expectedScore = 15 + 25 * (1 * 1.5) // 52.5
   console.log("More than 26 transactions. Expected tx points: "+ expectedScore)
-  for(let i = 0; i < 40; i++) {
+  for(let i = 0; i < 1; i++) {
```
This is weird code - i = 0 and < 1?? That's not an iteration ;)

IanMendozaJaimes replied:
Yes, that's right, my mistake sorry

n13 commented on test/harvest.test.js:
```diff
@@ -504,45 +528,48 @@ describe("harvest transaction score", async assert => {
   await checkScores([10, 16], [0, 50], "2 reputation, 2 tx", "0, 50 score")
 
   let expectedScore = 15 + 25 * (1 * 1.5) // 52.5
   console.log("More than 26 transactions. Expected tx points: "+ expectedScore)
```
comment needs to change

IanMendozaJaimes replied:
Sorry, I think I don't understand, do you mean this part: `'0'+memoprefix`?

n13 commented:
Updated conflicts

@IanMendozaJaimes 

This action is used but was removed from history

ACTION organization::scoreorgs(name next) {
    require_auth(get_self());

    auto itr = next == ""_n ? organizations.begin() : organizations.lower_bound(next.value);
    if(itr != organizations.end()) {
        action(
            permission_level{contracts::history, "active"_n},
            contracts::history, "orgtxpoints"_n,
            std::make_tuple(itr -> org_name)
        ).send();
        itr++;
        if (itr != organizations.end()) {
            action next_execution(
                permission_level{get_self(), "active"_n},
                get_self(),
                "scoreorgs"_n,
                std::make_tuple(itr->org_name)
            );
            transaction tx;
            tx.actions.emplace_back(next_execution);
            tx.delay_sec = 1; // change this to 1 to test
            tx.send(next.value+2, _self);
        }
    }
}

n13 reviewed (changes requested):
Need to fix call to "orgtxpoints" in history - call equivalent method?

IanMendozaJaimes commented:
Hello, I added all the changes requested. 
In the case of the action `scoreorgs`, I changed the action it calls, but `scoreorgs` is no longer needed, as all the transaction points calculations are done in harvest in the action `calctrxpts`. I forgot to remove the `scoreorgs` action sorry. In this commit I left it in case it is needed for other reason, but I removed it from scheduler's reset action.
